### ISSUE TITLE
A consumed otp becomes valid again within the drift period

### ISSUE THREAD

**Issue by mnipper** [tinfoil/devise-two-factor#106]
If an otp is consumed, it becomes invalid during the `otp.interval` of the instantiated `ROTP::TOTP` object in TwoFactorAuthenticatable.  `otp.interval` is currently always 30 seconds, since that is the default: https://github.com/mdp/rotp/blob/v3.3.0/lib/rotp/totp.rb#L10

However, during the drift period specified by `otp_allowed_drift`, the otp again becomes valid.

I believe this is unexpected behavior, and the intention is for an otp to not be reusable within the drift period after it has been consumed.

Here are some specs dropped in [here](https://github.com/tinfoil/devise-two-factor/blob/v3.0.0/lib/devise_two_factor/spec_helpers/two_factor_authenticatable_shared_examples.rb#L67) which currently fail:
```
      context 'given a valid OTP used multiple times within the allowed drift' do
        let(:consumed_otp) { ROTP::TOTP.new(otp_secret).at(Time.now) }

        before do
          subject.validate_and_consume_otp!(consumed_otp)
        end

        context 'after the otp interval' do
          before do
            Timecop.travel(subject.otp.interval.seconds.from_now)
          end

          # This currently fails!
          it 'fails to validate' do
            expect(subject.validate_and_consume_otp!(consumed_otp)).to be false
          end
        end
      end
```

**Comment by Edwardzyc** [comment]
Can we close this issue because of the merged PR?

**Comment by sadgb** [comment]
not fixed in master

**Comment by alexanderheckel** [comment]
Hi, any plans to merge / fix the master branch?

**Comment by oliverrauscher** [comment]
Hello! I really appreciate this great gem! 

Is there any chance that this critical fix will get merged into the master branch soon?

**Comment by alexanderheckel** [comment]
Hi, any updates on this?

**Comment by alexanderheckel** [comment]
Hi, any updates on this?

**Comment by eddyh** [comment]
The issue can be mitigated by disallowing OTP drift. To do so, add `Devise.otp_allowed_drift = 0` to your _devise.rb_ file.

I'm hoping this issue will be patched soon.

**Comment by bsedat** [comment]
Thank you, this is fixed now in the published version 4.0.2.

**Comment by codehelp** [comment]
Hi,
https://github.com/tinfoil/devise-two-factor/pull/108 remains open as a fix for this issue, so could you clarify please if 64576bb9e7d29800c5f92bb86fb6ecff91ad6105 is the fixing commit for this issue, as released in 4.0.2, and therefore for https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-43177 as well?

Thanks.
